File: WebHuntiOS/WebHunt/Networking/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case serverError([String])
    case decodingError(Error)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid URL"
        case .invalidResponse: return "Invalid server response"
        case .serverError(let messages): return messages.joined(separator: ", ")
        case .decodingError(let error): return "Failed to parse response: \(error.localizedDescription)"
        case .networkError(let error): return error.localizedDescription
        }
    }
}

actor APIClient {
    static let shared = APIClient()

    // Update this to point to your Rails backend
    #if targetEnvironment(simulator)
    private let baseURL = "http://localhost:3000"
    #else
    private let baseURL = "https://webhunt.dev"
    #endif

    private let session: URLSession
    private var csrfToken: String?

    init() {
        let config = URLSessionConfiguration.default
        config.httpCookieAcceptPolicy = .always
        config.httpShouldSetCookies = true
        config.httpCookieStorage = HTTPCookieStorage.shared
        self.session = URLSession(configuration: config)
    }

    /// Fetch the CSRF token from the Rails root page meta tag.
    /// Called automatically before POST/PATCH/DELETE requests.
    private func ensureCSRFToken() async throws {
        if csrfToken != nil { return }
        guard let url = URL(string: baseURL + "/") else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.setValue("text/html", forHTTPHeaderField: "Accept")
        let (data, _) = try await session.data(for: request)
        if let html = String(data: data, encoding: .utf8),
           let range = html.range(of: "csrf-token\" content=\""),
           let endRange = html[range.upperBound...].range(of: "\"") {
            csrfToken = String(html[range.upperBound..<endRange.lowerBound])
        }
    }

    // MARK: - Session / Auth

    func login(username: String, password: String) async throws -> SessionResponse {
        let body: [String: Any] = [
            "user": [
                "username": username,
                "password": password
            ]
        ]
        let data = try await post(path: "/api/session", body: body)
        return try JSONDecoder().decode(SessionResponse.self, from: data)
    }

    func signup(username: String, email: String, password: String) async throws -> SessionResponse {
        let body: [String: Any] = [
            "user": [
                "username": username,
                "email": email,
                "password": password
            ]
        ]
        let data = try await post(path: "/api/users", body: body)
        return try JSONDecoder().decode(SessionResponse.self, from: data)
    }

    func logout() async throws {
        _ = try await delete(path: "/api/session")
        csrfToken = nil
    }

    // MARK: - Products

    func fetchProducts() async throws -> ProductsIndexResponse {
        let data = try await get(path: "/api/products")
        return try JSONDecoder().decode(ProductsIndexResponse.self, from: data)
    }

    func fetchProduct(id: Int) async throws -> Product {
        let data = try await get(path: "/api/products/\(id)")
        return try JSONDecoder().decode(Product.self, from: data)
    }

    func createProduct(title: String, tagline: String, website: String, description: String, hunterId: Int, logoData: Data?, logoFilename: String?) async throws -> Product {
        var fields: [String: String] = [
            "product[title]": title,
            "product[tagline]": tagline,
            "product[website]": website,
            "product[description]": description,
            "product[hunter_id]": String(hunterId),
            "product[launch_date]": ISO8601DateFormatter().string(from: Date()),
            "product[status]": "true"
        ]

        var fileFields: [(fieldName: String, data: Data, filename: String, mimeType: String)] = []
        if let logoData = logoData, let filename = logoFilename {
            fileFields.append((
                fieldName: "product[logo]",
                data: logoData,
                filename: filename,
                mimeType: "image/jpeg"
            ))
        }

        let data = try await multipartPost(path: "/api/products", fields: fields, files: fileFields)
        return try JSONDecoder().decode(Product.self, from: data)
    }

    // MARK: - Users

    func fetchUser(id: Int) async throws -> User {
        let data = try await get(path: "/api/users/\(id)")
        return try JSONDecoder().decode(User.self, from: data)
    }

    func fetchUserByUsername(username: String) async throws -> User {
        let data = try await get(path: "/api/username/\(username)")
        return try JSONDecoder().decode(User.self, from: data)
    }

    // MARK: - Reviews

    func fetchReviews(productId: Int) async throws -> [Review] {
        let data = try await get(path: "/api/products/\(productId)/reviews")
        let response = try JSONDecoder().decode(ReviewsIndexResponse.self, from: data)
        return Array(response.reviews.values).sorted { $0.id < $1.id }
    }

    func createReview(productId: Int, reviewerId: Int, body: String, parentReviewId: Int? = nil) async throws -> Review {
        var reviewBody: [String: Any] = [
            "reviewer_id": reviewerId,
            "product_id": productId,
            "body": body
        ]
        if let parentId = parentReviewId {
            reviewBody["parent_review_id"] = parentId
        }
        let body: [String: Any] = ["review": reviewBody]
        let data = try await post(path: "/api/reviews", body: body)
        return try JSONDecoder().decode(Review.self, from: data)
    }

    // MARK: - Upvotes

    func toggleUpvote(upvoteableType: String, upvoteableId: Int, userId: Int) async throws -> Bool {
        let body: [String: Any] = [
            "vote": [
                "upvoteable_type": upvoteableType,
                "upvoteable_id": upvoteableId,
                "user_id": userId
            ]
        ]
        let data = try await post(path: "/api/upvotes", body: body)
        return try JSONDecoder().decode(Bool.self, from: data)
    }

    // MARK: - HTTP Methods

    private func get(path: String) async throws -> Data {
        guard let url = URL(string: baseURL + path) else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        try validateResponse(response, data: data)
        return data
    }

    private func post(path: String, body: [String: Any]) async throws -> Data {
        try await ensureCSRFToken()
        guard let url = URL(string: baseURL + path) else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token = csrfToken {
            request.setValue(token, forHTTPHeaderField: "X-CSRF-Token")
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        try validateResponse(response, data: data)
        return data
    }

    private func delete(path: String) async throws -> Data {
        try await ensureCSRFToken()
        guard let url = URL(string: baseURL + path) else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token = csrfToken {
            request.setValue(token, forHTTPHeaderField: "X-CSRF-Token")
        }

        let (data, response) = try await session.data(for: request)
        try validateResponse(response, data: data)
        return data
    }

    private func multipartPost(path: String, fields: [String: String], files: [(fieldName: String, data: Data, filename: String, mimeType: String)]) async throws -> Data {
        try await ensureCSRFToken()
        guard let url = URL(string: baseURL + path) else {
            throw APIError.invalidURL
        }

        let boundary = UUID().uuidString
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token = csrfToken {
            request.setValue(token, forHTTPHeaderField: "X-CSRF-Token")
        }

        var bodyData = Data()

        for (key, value) in fields {
            bodyData.append("--\(boundary)\r\n".data(using: .utf8)!)
            bodyData.append("Content-Disposition: form-data; name=\"\(key)\"\r\n\r\n".data(using: .utf8)!)
            bodyData.append("\(value)\r\n".data(using: .utf8)!)
        }

        for file in files {
            bodyData.append("--\(boundary)\r\n".data(using: .utf8)!)
            bodyData.append("Content-Disposition: form-data; name=\"\(file.fieldName)\"; filename=\"\(file.filename)\"\r\n".data(using: .utf8)!)
            bodyData.append("Content-Type: \(file.mimeType)\r\n\r\n".data(using: .utf8)!)
            bodyData.append(file.data)
            bodyData.append("\r\n".data(using: .utf8)!)
        }

        bodyData.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = bodyData

        let (data, response) = try await session.data(for: request)
        try validateResponse(response, data: data)
        return data
    }

    private func validateResponse(_ response: URLResponse, data: Data) throws {
        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if let messages = try? JSONDecoder().decode([String].self, from: data) {
                throw APIError.serverError(messages)
            }
            throw APIError.serverError(["Request failed with status \(httpResponse.statusCode)"])
        }
    }
}

File: WebHuntAndroid/app/src/main/java/dev/webhunt/android/data/network/ApiClient.kt
package dev.webhunt.android.data.network

import dev.webhunt.android.BuildConfig
import dev.webhunt.android.data.model.Product
import dev.webhunt.android.data.model.ProductsIndexResponse
import dev.webhunt.android.data.model.Review
import dev.webhunt.android.data.model.ReviewsIndexResponse
import dev.webhunt.android.data.model.SessionResponse
import dev.webhunt.android.data.model.User
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class ApiClient(
    private val apiService: ApiService,
    private val csrfInterceptor: CsrfInterceptor,
    private val cookieJar: CookieJarImpl
) {
    private val baseUrl: String get() = BuildConfig.BASE_URL

    // MARK: - Auth

    suspend fun login(username: String, password: String): SessionResponse {
        val body = mapOf("user" to mapOf("username" to username, "password" to password))
        val response = apiService.login(body)
        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    suspend fun signup(username: String, email: String, password: String): SessionResponse {
        val body = mapOf("user" to mapOf("username" to username, "email" to email, "password" to password))
        val response = apiService.signup(body)
        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    suspend fun logout() {
        apiService.logout()
        csrfInterceptor.clearToken()
    }

    // MARK: - Products

    suspend fun fetchProducts(): ProductsIndexResponse {
        val response = apiService.fetchProducts()
        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    suspend fun fetchProduct(id: Int): Product {
        val response = apiService.fetchProduct(id)
        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    suspend fun createProduct(
        title: String,
        tagline: String,
        website: String,
        description: String,
        hunterId: Int,
        logoBytes: ByteArray?,
        logoFilename: String?
    ): Product {
        val textType = "text/plain".toMediaType()
        val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.US)

        val logoPart = if (logoBytes != null && logoFilename != null) {
            val requestBody = logoBytes.toRequestBody("image/jpeg".toMediaType())
            MultipartBody.Part.createFormData("product[logo]", logoFilename, requestBody)
        } else null

        val response = apiService.createProduct(
            title = title.toRequestBody(textType),
            tagline = tagline.toRequestBody(textType),
            website = website.toRequestBody(textType),
            description = description.toRequestBody(textType),
            hunterId = hunterId.toString().toRequestBody(textType),
            launchDate = dateFormat.format(Date()).toRequestBody(textType),
            status = "true".toRequestBody(textType),
            logo = logoPart
        )

        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    // MARK: - Users

    suspend fun fetchUser(id: Int): User {
        val response = apiService.fetchUser(id)
        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    suspend fun fetchUserByUsername(username: String): User {
        val response = apiService.fetchUserByUsername(username)
        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    // MARK: - Reviews

    suspend fun fetchReviews(productId: Int): List<Review> {
        val response = apiService.fetchReviews(productId)
        if (response.isSuccessful) {
            return response.body()!!.reviews.values.sortedBy { it.id }
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    suspend fun createReview(
        productId: Int,
        reviewerId: Int,
        body: String,
        parentReviewId: Int? = null
    ): Review {
        val reviewBody = mutableMapOf<String, Any>(
            "reviewer_id" to reviewerId,
            "product_id" to productId,
            "body" to body
        )
        if (parentReviewId != null) {
            reviewBody["parent_review_id"] = parentReviewId
        }
        val requestBody = mapOf("review" to reviewBody)
        val response = apiService.createReview(requestBody)
        if (response.isSuccessful) {
            return response.body()!!
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    // MARK: - Upvotes

    suspend fun toggleUpvote(upvoteableType: String, upvoteableId: Int, userId: Int): Boolean {
        val body = mapOf(
            "vote" to mapOf<String, Any>(
                "upvoteable_type" to upvoteableType,
                "upvoteable_id" to upvoteableId,
                "user_id" to userId
            )
        )
        val response = apiService.toggleUpvote(body)
        if (response.isSuccessful) {
            return response.body() ?: true
        }
        throw ApiException(parseError(response.errorBody()?.string()))
    }

    // MARK: - Helpers

    fun resolveUrl(path: String?): String? {
        if (path == null) return null
        if (path.startsWith("http")) return path
        return "$baseUrl$path"
    }

    private fun parseError(errorBody: String?): String {
        if (errorBody == null) return "Unknown error"
        // Try to parse as JSON array of strings: ["error message"]
        return try {
            val trimmed = errorBody.trim()
            if (trimmed.startsWith("[")) {
                trimmed.removeSurrounding("[", "]")
                    .split(",")
                    .joinToString(", ") { it.trim().removeSurrounding("\"") }
            } else {
                trimmed
            }
        } catch (_: Exception) {
            errorBody
        }
    }
}

class ApiException(message: String) : Exception(message)
